### DISCUSSION TITLE
Reading from wholeBodyDynamicsTree ports

### DISCUSSION THREAD

**Discussion by KevinAStein** [robotology/community#127] [Q&A] [answered]
Hi,
I want to read force-torque and joint sensors from iCub to estimate ZMP and COM position and velocity using wholeBodyDynamicsTree. Since the readings are quite noisy I tend to use mean values. The results are okay reading every port approximately 50 times. This takes quite long (~200ms).

Is there a way to parallelize this? In what rate is the sensor data read and streamed?

thanks,
Kevin

@hu-yue

**Comment by apaikan** [comment]
@KevinAStein  What do you exactly mean with paralleling it? 

Anyway, a quick method to access the port data streaming frequency is using yarp read along with the `envelope` option and then check the time-stamp differences printed in the second column: 

```
$ yarp read ... /portname envelope 
```

**Comment by traversaro** [comment, marked as answer]
Hi @KevinAStein,

External forces in `wholeBodyDynamicsTree` are published on the ports every period of `wholeBodyDynamicsTree`, that is by default set to `10 ms`. You can modify it with the `rate` period (sorry for the misleading name, I need to fix it). However, by decreasing the period you would probably get the same estimates streamed for multiple times, as most of the sensors of the robot publish data with a period of `~10 ms`. 

Nonetheless, I suggest that you use the low-pass filters embedded in `wholeBodyDynamicsTree`  (see how to set their parameters with `wholeBodyDynamicsTree --help`) to reject noise in the estimation, or use some "classical" low pass filter on your side, and you will likely need less than 50 samples to get a meaningful estimate. :)

**Comment by pattacini** [comment]
@KevinAStein have the recommendations given above been useful for you?

**Comment by KevinAStein** [comment]
thanks for your help. Using the already implemented filter made things a lot easier.
